stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java

```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
+                                        timestamp != null ? createTimeStruct(timestamp.getTime()) : null);
+                            } else if (isNillableType(structFields, index - 1, TypeTags.INT_TAG)) {
+                                bStruct.setRefField(++longRegIndex,
+                                        timestamp != null ? new BInteger(timestamp.getTime()) : null);
+                            } else {
+                                handleMismatchingFieldAssignment();
+                            }
+                        } else {
+                            if (timestamp != null) {
+                                if (fieldTypeTag == TypeTags.STRING_TAG) {
+                                    String timestmpValue = SQLDatasourceUtils.getString(timestamp);
+                                    bStruct.setStringField(++stringRegIndex, timestmpValue);
+                                } else if (fieldTypeTag == TypeTags.STRUCT_TAG) {
+                                    bStruct.setRefField(++refRegIndex, createTimeStruct(timestamp.getTime()));
+                                } else if (fieldTypeTag == TypeTags.INT_TAG) {
+                                    bStruct.setIntField(++longRegIndex, timestamp.getTime());
+                                }
+                            } else {
+                                handleNilToNonNilableFieldAssignment();
+                            }
                         }
                         break;
                     case Types.ROWID:
-                        BValue strValue = new BString(new String(rs.getRowId(index).getBytes(), "UTF-8"));
-                        bStruct.setStringField(++stringRegIndex, strValue.stringValue());
+                        sValue = new String(rs.getRowId(index).getBytes(), "UTF-8");
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = new BString(sValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.STRING_TAG);
+                        } else {
+                            bStruct.setStringField(++stringRegIndex, sValue);
+                        }
                         break;
                     case Types.TINYINT:
                     case Types.SMALLINT:
                         long iValue = rs.getInt(index);
-                        bStruct.setIntField(++longRegIndex, iValue);
+                        isOriginalValueNull = rs.wasNull();
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = isOriginalValueNull ? null : new BInteger(iValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.INT_TAG);
+                        } else {
+                            if (!isOriginalValueNull) {
```

Comment:
What if we change the if else cases as follows by avoiding unnecessary ! here
if (isOriginalValueNull) {
handleNilToNonNilableFieldAssignment();
} else {
//set the value.
}